src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadManagerImpl.java

```diff
@@ -0,0 +1,654 @@
+        Set<OffsetDateTime> pushedDateTimes = getDateTimeValuesFromDeviceData( data );
+        if ( pushedDateTimes.size() == 0 ) {
+            return null;
+        }
+
+        String firstDateTime = pushedDateTimes
+                .stream()
+                .min( OffsetDateTime::compareTo )
+                .orElse( null )
+                .toString();
+
+        String lastDateTime = pushedDateTimes
+                .stream()
+                .max( OffsetDateTime::compareTo )
+                .orElse( null )
+                .toString();
+
+        Set<Object> uniqueDates = pushedDateTimes
+                .stream()
+                .map( dt -> dt
+                        .truncatedTo( ChronoUnit.DAYS )
+                        .format( DateTimeFormatter.ISO_DATE_TIME ) )
+                .collect( Collectors.toSet() );
+
+        Map<UUID, Set<Object>> metadataEntityData = getMetadataEntity( participantEKID );
+        EntityKey metadataEK = getMetadataEntityKey( metadataESID, metadataEntityData );
+
+        // verify if there is already an entry of metadata for participant
+        // error means there is no metadata yet.
+        UUID metadataEntityKeyId = integrationApi.getEntityKeyIds( ImmutableSet.of( metadataEK ) ).iterator().next();
+        Map<FullQualifiedName, Set<Object>> entity = new HashMap<>();
+        try {
+            entity = dataApi.getEntity( metadataESID, metadataEntityKeyId );
+        } catch ( Exception exception ) {
+            logger.error(
+                    "failure while getting metadata entity {} - study = {} participant = {}",
+                    metadataEntityKeyId,
+                    studyId,
+                    participantId,
+                    exception
+            );
+        }
+        metadataEntityData.put( commonTasksManager.getPropertyTypeId( START_DATE_TIME_FQN ),
+                entity.getOrDefault( START_DATE_TIME_FQN, Set.of( firstDateTime ) ) );
+        metadataEntityData.put( commonTasksManager.getPropertyTypeId( END_DATE_TIME_FQN ), Set.of( lastDateTime ) );
+        uniqueDates.addAll( entity.getOrDefault( RECORDED_DATE_TIME_FQN, Set.of() ) );
+        metadataEntityData.put( commonTasksManager.getPropertyTypeId( RECORDED_DATE_TIME_FQN ), uniqueDates );
+
+        entitiesByEntityKey.put( metadataEK, metadataEntityData );
+
+        Map<UUID, Set<Object>> hasEntityData = getHasEntity( firstDateTime );
+        EntityKey hasEK = getHasEntityKey( hasESID, hasEntityData );
+        entitiesByEntityKey.put( hasEK, hasEntityData );
+
+        EntityKey participantEK = getParticipantEntityKey( participantESID, participantId );
+
+        // association: participant  => has => metadata
+        edgesByEntityKey.add( Triple.of( participantEK, hasEK, metadataEK ) );
+
+        return Pair.of( entitiesByEntityKey, edgesByEntityKey );
+    }
+
+    private String getFirstValueOrNullByUUID( SetMultimap<UUID, Object> entity, FullQualifiedName fqn ) {
```

Comment:
you shouldn't need to state `byUUID` explicitly here -- you could have two methods both called `getFirstValueOrNull`, one which takes a map with UUID as key and one which takes a map with FQN as key, and the class would compile since the param type would imply which function would be invoked